### PR TITLE
[UAR-276] review trusts former beneficial owners

### PR DESCRIPTION
**Pull request by corybudd-ch**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-276

### Change description
Summary page for former BOs updated.
New page to add former BO.

Screenshot 2023-09-22 at 15 15 30

Screenshot 2023-09-22 at 10 25 18

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/utils/update/review_trusts.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      case TrusteeType.INDIVIDUAL:
+        return (trust?.INDIVIDUALS ?? []).findIndex(trustee => trustee.id === trusteeId);
+      case TrusteeType.LEGAL_ENTITY:
+        return (trust?.CORPORATES ?? []).findIndex(trustee => trustee.id === trusteeId);
+      default:
+        return -1;
+  }
+};
+
+export const getTrustee = (appData: ApplicationData, trust: Trust, trusteeId: string, trusteeType: string) => {
```
Comment (conuirchgov) [inline src/utils/update/review_trusts.ts:24]: appData doesn't appear to be used here :)

Comment (corybudd-ch) [reply src/utils/update/review_trusts.ts:24]: Good spot. My VScode not bringing it to my attention at all

File: src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts [hunk 2]
```diff
@@ -2,68 +2,130 @@ import { NextFunction, Request, Response } from 'express';
...
+
+    const session = req.session as Session;
+    const appData = getApplicationData(session);
+    const trusteeId = req.params[ROUTE_PARAM_TRUSTEE_ID];
+    const formData: TrustHistoricalBeneficialOwnerForm = req.body as TrustHistoricalBeneficialOwnerForm;
+
+    // check for form validation errors
+    const errorList = validationResult(req);
+    if (!errorList.isEmpty()) {
+      const pageProperties = getPageProperties({} as Trust, formData, formatValidationError(errorList.array()));
```
[the remark this answers is not in the document]
Comment (corybudd-ch) [reply src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts:55]: You are right ... I have no idea why I passed in {} haha ..

File: src/utils/update/review_trusts.ts (hunk 1 above)
Comment (corybudd-ch) [reply src/utils/update/review_trusts.ts]: fixed

File: src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (corybudd-ch) [reply src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts]: fixed

File: src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      trustId: trust.trust_id,
+      trustName: trust.trust_name
+    },
+    trusteeType: TrusteeType,
+  },
+  formData,
+  errors
+});
+
+const mapFormerBOFromModelToForm = (
```
Comment (stephanie-ch) [inline src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts:94]: We can save on this duplication actually, check Conuir's PR

File: src/controllers/update/update.manage.trusts.review.former.bo.controller.ts [hunk 4]
```diff
@@ -2,30 +2,59 @@ import { NextFunction, Request, Response } from 'express';
...
   try {
     logger.debugRequest(req, `${req.method} ${req.route.path}`);
 
-    return res.redirect(UPDATE_MANAGE_TRUSTS_REVIEW_INDIVIDUALS_URL);
+    if (req.body.addFormerBo) {
+      return res.redirect(UPDATE_MANAGE_TRUSTS_TELL_US_ABOUT_THE_FORMER_BO_URL);
+    }
+
+    const appData = getApplicationData(req.session);
+    setFormerBOsAsReviewed(appData);
```
[the remark this answers is not in the document]
Comment (corybudd-ch) [reply src/controllers/update/update.manage.trusts.review.former.bo.controller.ts]: done, thanks!

File: src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts (hunk 3 above)
Comment (corybudd-ch) [reply src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts:94]: done, thanks!

**Review by joshuahigginsonCH — APPROVED** [review]

**Review by stephanie-ch — APPROVED** [review]

File: src/controllers/update/update.manage.trusts.review.former.bo.controller.ts [hunk 5]
```diff
@@ -2,29 +2,50 @@ import { NextFunction, Request, Response } from 'express';
...
+import { Trust } from '../../model/trust.model';
+import { Session } from '@companieshouse/node-session-handler';
+import { TrusteeType } from '../../model/trustee.type.model';
 
 export const get = (req: Request, res: Response, next: NextFunction) => {
   try {
     logger.debugRequest(req, `${req.method} ${req.route.path}`);
 
+    const appData = getApplicationData(req.session as Session);
+    const trust = getTrustInReview(appData) as Trust;
```
Comment (AndreKramer7654) [inline src/controllers/update/update.manage.trusts.review.former.bo.controller.ts:22]: Should not need to use "as Trust"? Or should be checking for undefined which means nothing to review?

File: src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import { saveAndContinue } from '../../utils/save.and.continue';
+import { mapBeneficialOwnerToSession, mapFormerTrusteeFromSessionToPage } from '../../utils/trust/historical.beneficial.owner.mapper';
+import { getTrustInReview, getTrustee, getTrusteeIndex } from '../../utils/update/review_trusts';
+
+export const get = (req: Request, res: Response, next: NextFunction) => {
+  try {
+    logger.debugRequest(req, `${req.method} ${req.route.path}`);
+
+    const appData = getApplicationData(req.session);
+    const trust = getTrustInReview(appData) as Trust;
```
Comment (AndreKramer7654) [inline src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts:25]: As comment above - may be better to redirect or error if no trust in review.

Comment (corybudd-ch) [reply src/controllers/update/update.manage.trusts.tell.us.about.the.former.bo.controller.ts:25]: Same as above, middleware will catch and redirect if no trust in review

**Pull request merged**